zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/util/ZkPathRecursiveWatcherTrie.java

```diff
@@ -0,0 +1,271 @@
+     *
+     * @return true if this node is a lead node; otherwise false
+     */
+    boolean isLeafNode() {
+      return children.isEmpty();
+    }
+
+    /**
+     * Get the set of RecursivePersistWatcherListener
+     * Returns an empty set if no listener is registered on the path
+     * @return
+     */
+    @VisibleForTesting
+    Set<RecursivePersistListener> getRecursiveListeners() {
+      return recursiveListeners;
+    }
+
+    @Override
+    public String toString() {
+      return "TrieNode [name=" + value + ", children=" + children.keySet() + "]";
+    }
+  }
+
+  /**
+   * Construct a new PathTrie with a root node.
+   */
+  public ZkPathRecursiveWatcherTrie() {
+    this.rootNode = new TrieNode(null, "/");
+  }
+
+  /**
+   * Add a path to the path trie. All paths are relative to the root node.
+   *
+   * @param path the path to add RecursivePersistListener
+   * @param listener the RecursivePersistListener to be added
+   */
+  public void addRecursiveListener(final String path, RecursivePersistListener listener) {
+    Objects.requireNonNull(path, "Path cannot be null");
+
+    if (path.length() == 0) {
+      throw new IllegalArgumentException("Invalid path: " + path);
+    }
+    final String[] pathComponents = split(path);
+
+    writeLock.lock();
+    try {
+      TrieNode parent = rootNode;
+      for (final String part : pathComponents) {
+        TrieNode child = parent.getChild(part);
+        if (child == null) {
+          child = new TrieNode(parent, part);
+          parent.addChild(part, child);
+        }
+        parent = child;
+      }
+      parent.recursiveListeners.add(listener);
+    } finally {
+      writeLock.unlock();
+    }
+  }
+
+  /**
+   * Removing a RecursivePersistWatcherListener on a path.
+   *
+   * Delete a path from the nearest trie node to current node if this is the only listener and there
+   * is no child on the trie node.
+   *
+   * @param path the of the lister registered
+   * @param listener the RecursivePersistListener to be removed
+   */
+  public void removeRecursiveListener(final String path, RecursivePersistListener listener) {
+    Objects.requireNonNull(path, "Path cannot be null");
+
+    if (path.length() == 0) {
+      throw new IllegalArgumentException("Invalid path: " + path);
+    }
+    final String[] pathComponents = split(path);
+
+    writeLock.lock();
+    try {
+      TrieNode parent = rootNode;
+      for (final String part : pathComponents) {
+        if (parent.getChild(part) == null) {
+          // the path does not exist
+          return;
+        }
+        parent = parent.getChild(part);
+      }
+
+      TrieNode cur = parent;
+      cur.getRecursiveListeners().remove(listener);
+      // if cur is leaf node, remove the node and the path up to root until and parent node
+      // 1. has other child, or has watcher on that path
+      // remove the path if needed
+      while (cur != rootNode && (cur.children.size() == 0
```

Comment:
Why not parallel delete? If we decide to delete the entire sub trie, then the each of branch of trie should be independent and can be parallel processing.

Reply:
This delete is from leaf to root. I am not sure if there is anything to be parallel processed..?

Reply:
Update as suggested. No need to delete from leaf to root. Will add this API back in following change.